File: web/lib/types/self-study.ts
/** Bağımsız çalışma kayıtları — API v2 tipleri (backend schemas/self_study.py aynası). */

export type SelfStudySource = "student" | "coach";
export type SelfStudyStatus = "pending" | "approved" | "rejected";

export interface SelfStudyEntryItem {
  id: number;
  student_book_id: number;
  book_id: number;
  book_name: string;
  subject_name: string;
  section_id: number;
  section_label: string;
  test_count: number;
  applied_count: number;
  source: SelfStudySource;
  source_label: string;
  status: SelfStudyStatus;
  status_label: string;
  note: string | null;
  period_start: string | null;
  period_end: string | null;
  created_by_name: string | null;
  created_at: string;
  reviewed_at: string | null;
  review_note: string | null;
}

export interface SelfStudyListResponse {
  items: SelfStudyEntryItem[];
  pending_count: number;
}

export interface SelfStudyCreateItem {
  student_book_id: number;
  section_id: number;
  test_count: number;
}

export interface SelfStudyCreateBody {
  items: SelfStudyCreateItem[];
  note?: string | null;
  period_start?: string | null;
  period_end?: string | null;
}

export interface SelfStudySkippedItem {
  section_id: number;
  section_label: string;
  reason: string;
}

export interface SelfStudyCreateResult {
  created: SelfStudyEntryItem[];
  skipped: SelfStudySkippedItem[];
  applied_total: number;
  pending_total: number;
}

export interface SelfStudyReviewBody {
  approve: boolean;
  review_note?: string | null;
}

export interface SelfStudyDeleteResult {
  deleted_id: number;
  reverted_count: number;
}

export interface SelfStudyOptionSection {
  section_id: number;
  label: string;
  test_count: number;
  completed_count: number;
  reserved_count: number;
  remaining: number;
}

export interface SelfStudyOptionBook {
  student_book_id: number;
  book_id: number;
  book_name: string;
  subject_name: string;
  book_type_label: string;
  sections: SelfStudyOptionSection[];
}

export interface SelfStudyOptionsResponse {
  books: SelfStudyOptionBook[];
}

File: app/routes/api_v2/schemas/self_study.py
"""Bağımsız çalışma kayıtları — API v2 şemaları (öğrenci + koç ortak)."""
from __future__ import annotations

from datetime import date

from pydantic import BaseModel, Field


class SelfStudyEntryItem(BaseModel):
    id: int
    student_book_id: int
    book_id: int
    book_name: str
    subject_name: str
    section_id: int
    section_label: str
    test_count: int
    applied_count: int
    source: str                  # student | coach
    source_label: str
    status: str                  # pending | approved | rejected
    status_label: str
    note: str | None
    period_start: str | None     # ISO date
    period_end: str | None
    created_by_name: str | None
    created_at: str              # ISO
    reviewed_at: str | None
    review_note: str | None


class SelfStudyListResponse(BaseModel):
    items: list[SelfStudyEntryItem]
    pending_count: int


class SelfStudyCreateItem(BaseModel):
    student_book_id: int
    section_id: int
    test_count: int = Field(ge=1, le=1000)


class SelfStudyCreateBody(BaseModel):
    items: list[SelfStudyCreateItem] = Field(min_length=1, max_length=100)
    note: str | None = Field(default=None, max_length=500)
    period_start: date | None = None
    period_end: date | None = None


class SelfStudySkippedItem(BaseModel):
    section_id: int
    section_label: str
    reason: str


class SelfStudyCreateResult(BaseModel):
    created: list[SelfStudyEntryItem]
    skipped: list[SelfStudySkippedItem]
    applied_total: int           # ilerlemeye gerçekten işlenen toplam test
    pending_total: int           # onay bekleyen toplam test (öğrenci beyanı)


class SelfStudyReviewBody(BaseModel):
    approve: bool
    review_note: str | None = Field(default=None, max_length=500)


class SelfStudyDeleteResult(BaseModel):
    deleted_id: int
    reverted_count: int          # ilerlemeden geri alınan test


# --- Öğrenci beyan dialogu için kitap/bölüm seçenekleri ---


class SelfStudyOptionSection(BaseModel):
    section_id: int
    label: str
    test_count: int
    completed_count: int
    reserved_count: int
    remaining: int


class SelfStudyOptionBook(BaseModel):
    student_book_id: int
    book_id: int
    book_name: str
    subject_name: str
    book_type_label: str
    sections: list[SelfStudyOptionSection]


class SelfStudyOptionsResponse(BaseModel):
    books: list[SelfStudyOptionBook]
